W-12450976 Update upgrade tool description (docs:  W-12481917) (#2452)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/mule-upgrade-tool.adoc

= Mule Upgrade Tool

> The Mule upgrade tool enables you to update or upgrade Mule 4 instances located in different folders.

Reviewer: Is the tool for upgrading standalone Mule instances that are hosted on-prem whether alone or in a clustered environment? Or can it upgrade in other hosted environments?

And what types of upgrades does it do? major version, minor version, patch version, or patch update version?

Author: Answered in a call

If the Runtime Manager agent is installed in your current Mule instance, the upgrade tool also updates the agent version as part of the upgrade process.

> The Mule upgrade tool supports upgrading clustered Mule instances by manually upgrading each node using the tool. See xref:release-notes::mule-runtime/updating-mule-4-versions.adoc#mulerunvers[Upgrading an On-Premises Mule Instance Managed Through Runtime Manager] for more information.

Reviewer: We are in discussion about if/why this sentence is needed on this page: 

> See xref:release-notes::mule-runtime/updating-mule-4-versions.adoc#mulerunvers[Upgrading an On-Premises Mule Instance Managed Through Runtime Manager] for more information.

Author: Removed

Reviewer 2: Adding the following to crosslink documents:
"Though highly recommended, using the tool is not strictly necessary to upgrade your current Mule instance. For a completely manual upgrade, see xref:release-notes::mule-runtime/updating-mule-4-versions.adoc#mulerunvers[Upgrading an On-Premises Mule Instance Managed Through Runtime Manager]."

= Mule Upgrade Tool

> *Latest version: 1.1.1*

Reviewer: ```suggestion
```

Reviewer: This won't be ready from the release and probably isn't needed because the latest version is clear from https://docs.mulesoft.com/release-notes/mule-upgrade-tool/mule-upgrade-tool, which is maintained as part of each version release.

Author: ok, I'll remove it

Author: done

> include::release-notes::partial$mule-upgrade-tool/mule-upgrade-tool-1.1.1.adoc[tag=intro]

Reviewer: This should be a general intro that is independent of versions. We don't want to maintain versioning here. 

If we're going to use an include, let's see if we can do this differently.

Reviewer: See PR https://github.com/mulesoft/docs-release-notes/pull/6060/files where we will need to sort this out.

> include::release-notes::partial$mule-upgrade-tool/mule-upgrade-tool-1.1.1.adoc[tag=intro]

Reviewer: It appears that Julio renamed this include to `mule-upgrade-tool/mule-upgrade-tool-1.0.0.adoc` in https://github.com/mulesoft/docs-release-notes/pull/6060/files

> include::release-notes::partial$mule-upgrade-tool/mule-upgrade-tool-1.1.1.adoc[tag=intro]

Reviewer: ```suggestion
include::release-notes::partial$mule-upgrade-tool/mule-upgrade-tool.adoc[tag=intro]
```

The Mule upgrade tool enables you to update or upgrade Mule 4 instances located in different folders.

If the Runtime Manager agent is installed in your current Mule instance, the upgrade tool also updates the agent version as part of the upgrade process.

> The Mule upgrade tool support upgrading clustered Mule instances by manually upgrading each node using the tool. For further information on how to proceed with the upgrade, please refer to section <<upgrade_clustered_mule_instances>>.

Reviewer 2: ```suggestion
The Mule upgrade tool supports upgrading clustered Mule instances by manually upgrading each node using the tool. See xref:release-notes::mule-runtime/updating-mule-4-versions.adoc#mulerunvers[Upgrading an On-Premises Mule Instance Managed Through Runtime Manager] for more information.
For additional information on how to proceed with the upgrade, see <<upgrade_clustered_mule_instances>>.
```

Author: Done

> The manual procedure you need to follow to do it manually is described in xref:release-notes::mule-runtime/updating-mule-4-versions.adoc#mulerunvers[Upgrading an On-Premises Mule Instance Managed Through Runtime Manager].

Reviewer 2: ```suggestion
```
I rephrased it above.

Author: done

= Mule Upgrade Tool

> include::release-notes::partial$mule-upgrade-tool/mule-upgrade-tool.adoc[tag=intro]

Reviewer: This says it's for upgrading 4.1.5 and later, not 4.1.1
https://github.com/mulesoft/docs-release-notes/blob/latest/modules/ROOT/pages/_partials/mule-upgrade-tool/mule-upgrade-tool.adoc

But this could be included in versions of the doc for Mule releases that have older versions of the tool.

Reviewer 2: Modifying it to include::release-notes::partial$mule-upgrade-tool/mule-upgrade-tool-1.1.0.adoc[tag=intro]

Perform a manual upgrade to ensure that the current and new Mule instances are in version 4.1.5 or later before attempting to upgrade using the Mule upgrade tool.

[[upgrade_clustered_mule_instances]]
== Upgrading Clustered Mule Instances

> The Mule upgrade tool does not automatically support upgrading clustered Mule instances. To upgrade clustered Mule instances, you have to manually upgrade each node using the tool.

Reviewer 2: ```suggestion
To upgrade clustered Mule instances, use the Mule upgrade tool to automate the upgrade process of your local Mule instances by manually upgrading each node.
```

What do you think about this, Jorge? I find it as a way to link what it says under the "mule upgrade tool" title and what is mentioned  in [Upgrading a Standalone Mule Instance](https://docs.mulesoft.com/release-notes/mule-runtime/updating-mule-4-versions#upgrading-a-standalone-mule-instance). Plus, we are adding a positive connotation.

Author: sounds good to me

Reviewer 2: Reworded.

=== Before You Begin

> Before starting your upgrade, ensure that you have:

Reviewer: "... ensure that you have:"
is followed by an imperative bullet point "* Copy ..." 

Should it be "* A copy of" or should "ensure that you have:" change to something else that flows into the imperative?

Reviewer 2: Modifying the first bullet.

> * The new Mule runtime (with the version to be upgraded to) and Mule upgrade tool instances copied to each node that must be upgraded, to ensure less downtime.

Reviewer: Can you get a suggestion for the wording of this bit: "The new Mule runtime (with the version to be upgraded to)"

Reviewer: I think we're saying that they need to have downloaded and unpacked the version of Mule to which they are upgrading. "with" could make it sound like we're adding something to Mule, and the "to be upgraded to" is understandable, but not sure if that follows our style guidelines.

Reviewer 3: I think it sounds odd to talk about the **Mule upgrade tool instance**, I'd suggest something like: 
_The new Mule runtime distribution/instance (target version) and the Mule upgrade tool copied to each node that you are upgrading, to ensure less downtime._
I think we call it "instance" when it is running, but not sure, maybe distribution sounds more like the executable?

Reviewer 4: Before starting the upgrade, copy the Mule upgrade tool and Mule runtime distribution to each node, to ensure less downtime.

Reviewer 4: Isn't it obvious that they need to copy the version of Mule that they're upgrading to?

Should "ensure less downtime" be "to reduce downtime"?

Reviewer 2: Thanks a lot for the contributions!
As I already have a "before starting the upgrade" section, and the sentence in discussion is a bullet of many others (located in the include), I will go with: 
_The new Mule runtime distribution (target version) and the Mule upgrade tool copied to each node that you are upgrading, to reduce downtime._

Perform a manual upgrade to ensure that the current and new Mule instances are in version 4.1.5 or later before attempting to upgrade using the Mule upgrade tool.

[[upgrade_clustered_mule_instances]]

> == Upgrading clustered Mule instances

Reviewer 2: ```suggestion
== Upgrading Clustered Mule Instances
```

Author: done

> Currently, the upgrade tool does not automatically support upgrading Mule clustered instances, and because of that you can find here a description of what to do to upgrade it using the Mule upgrade tool.

Reviewer 2: ```suggestion
The Mule upgrade tool does not automatically support upgrading clustered Mule instances. To upgrade clustered Mule instances, you have to manually upgrade each node using the tool.
```

Author: Done

=== Before You Begin

> To ensure less downtime, first copy the new instance of the Mule runtime and the Mule upgrade tool to each instance that will be upgraded. That way you can have everything you need to do the upgrade.

Reviewer 2: ```suggestion
Before starting your upgrade:
* Copy the new Mule runtime and Mule upgrade tool instances to each instance that is upgraded to ensure less downtime.
```

Author: Actually here we need to copy the new version and the tool to do the upgrade, and if we do it beforehand, we will have less downtime, that's the idea why we should do this in advance.

Author: I have rephrased it to

> * Copy the new Mule runtime (with the version to be upgraded to) and Mule upgrade tool instances to each node that must be upgraded, to ensure less downtime.

> Follow the same procedures as are described in the <<before_you_begin>> section.

Reviewer 2: ```suggestion
* Follow the procedures described in the above <<before_you_begin>> section.
```

Author: Done

=== Upgrade or Update Mule

The procedure you need to follow is the same as in the stand-alone instance upgrade.
Follow the same procedure described above, starting from the server node, and continuing with every other node till each of those are upgraded to the same version.

> This way your system will be properly updated and working as a cluster with the proper new version.

Reviewer 2: ```suggestion
To upgrade or update your clustered Mule instances:
. Follow the procedures as in the stand-alone instance upgrade described in the above <<upgrade_or_update_mule>> section. Start from the server node and continue with the other nodes until each node is upgraded to the same version. 
This way, your system will work as a cluster with the corresponding new versions. 
```
The 2 procedures mentioned are the same, right? The <<upgrade_or_update_mule>> from the tool? 
I will add the [[]] reference.

Author: Done

> === Rollback Mule

Reviewer 2: ```suggestion
=== Roll Back an Upgrade or Update
```
Comment:
`Roll back` is the verb form. 
`Rollback` is the noun form.

Author: Thanks for that, I appreciate it.

Author: Done

> The same thing that happens with the upgrade happen with the rollback procedure, every node must be rolled back and ends up having the same version. Because of that, you will need to roll back every node in the same order as the upgrade, first the server node, and then each individual node that is part of the cluster, making sure that, at the end, every node has the same version.

Reviewer 2: ```suggestion
Before starting an upgrade or update process, the Mule upgrade tool generates a backup of the clustered Mule instance you are upgrading or updating, which enables you to restore the upgraded or updated clustered Mule instance to its previous state. The backup is stored in the `$MULE_HOME/.upgrade-tool-backup` folder.
To perform a rollback:
. Follow the procedures described in the above <<roll_back_an_upgrade-or-update>> section. Every node must be rolled back in the same order as they were upgraded or updated starting from the server node and continuing with the other nodes that are part of the cluster. Once you finish, make sure that every node has the same version.
```
1. Added some info as per the <<roll_back_an_upgrade-or-update>> tool description considering that I want to modify the navigation tree as a next step to improve customer experience. Please check if it's OK. 
2. What do you mean by "same version"? Are we referring to the Original version?

I will add the [[]] reference.

Author: Done

Perform a manual upgrade to ensure that the current and new Mule instances are in version 4.1.5 or later before attempting to upgrade using the Mule upgrade tool.

[[upgrade_clustered_mule_instances]]
== Upgrading Clustered Mule Instances

The Mule upgrade tool does not automatically support upgrading clustered Mule instances. To upgrade clustered Mule instances, you have to manually upgrade each node using the tool.

> === Before You Begin

Reviewer: If there's another section in the same guide called Before You Begin, and then we point to that section from this Before You Begin section, I'm kind of wondering if we want to have a section with the same title here. Do we do that elsewhere in the docs?

Reviewer: See the suggestion about using includes in the comment below (under "Upgrade or Update Mule").

Reviewer 4: 100% we should avoid pointing people to different sections, even if the content is identical (it's hard for the user to keep track and doesn't provide a targeted experience). Definitely use includes and duplicate the relevant content in both places.

Author: changed

Reviewer 2: Creating includes.

Before starting your upgrade:

* Copy the new Mule runtime (with the version to be upgraded to) and Mule upgrade tool instances to each node that must be upgraded, to ensure less downtime.
* Follow the procedures described in the above <<before_you_begin>> section.

=== Upgrade or Update Mule

To upgrade or update your clustered Mule instances:

> Follow the procedures as in the stand-alone instance upgrade described in the above <<upgrade_or_update_mule>> section. Start from the server node and continue with the other nodes until each node is upgraded to the same version.

Reviewer: Regarding all the sections where we are pointing back to the same procedure above (saying, "Follow the procedures described in the above <<roll_back_an_upgrade-or-update>> section"), I chatted with Hanna who suggested using includes instead of links here. The idea was that here we are talking to people who are upgrading clusters. Above they are upgrading a single instance. Includes will avoid the need to click the links for the steps.

Reviewer 4: 100% we should avoid pointing people to different sections, even if the content is identical (it's hard for the user to keep track and doesn't provide a targeted experience). Definitely use includes and duplicate the relevant content in both places.

Author: I don't know about that. That's not my way of thinking, and programming, so I will have to trust in you about this, but is it not an easy thing for me to do it this way 🤣

Reviewer 2: Creating includes.
